pkgs/build-support/setup-hooks/make-binary-wrapper.sh

```diff
@@ -0,0 +1,198 @@
+    printf "\n%s" "$main"
+    printf "%s\n" "}"
+}
+
+addFlags() {
+    local result n flag flags
+    local var="argv_tmp"
+    flags=("$@")
+    for ((n = 0; n < ${#flags[*]}; n += 1)); do
+        flag=$(escapeStringLiteral "${flags[$n]}")
+        result="$result    $var[$((n+1))] = \"$flag\";"$'\n'
+    done
+    printf "    %s\n" "char **$var = malloc(sizeof(*$var) * ($((n+1)) + argc));"
+    printf "    %s\n" "$var[0] = argv[0];"
+    printf "%s" "$result"
+    printf "    %s\n" "for (int i = 1; i < argc; ++i) {"
+    printf "    %s\n" "    $var[$n + i] = argv[i];"
+    printf "    %s\n" "}"
+    printf "    %s\n" "$var[$n + argc] = NULL;"
+    printf "    %s\n" "argv = $var;"
+}
+
+# prefix ENV SEP VAL
+setEnvPrefix() {
+    local env=$(escapeStringLiteral "$1")
+    local sep=$(escapeStringLiteral "$2")
+    local val=$(escapeStringLiteral "$3")
+    printf "%s" "set_env_prefix(\"$env\", \"$sep\", \"$val\");"
+}
+
+# suffix ENV SEP VAL
+setEnvSuffix() {
+    local env=$(escapeStringLiteral "$1")
+    local sep=$(escapeStringLiteral "$2")
+    local val=$(escapeStringLiteral "$3")
+    printf "%s" "set_env_suffix(\"$env\", \"$sep\", \"$val\");"
+}
+
+# setEnv KEY VALUE
+setEnv() {
+    local key=$(escapeStringLiteral "$1")
+    local value=$(escapeStringLiteral "$2")
+    printf "%s" "putenv(\"$key=$value\");"
+}
+
+# setDefaultEnv KEY VALUE
+setDefaultEnv() {
+    local key=$(escapeStringLiteral "$1")
+    local value=$(escapeStringLiteral "$2")
+    printf "%s" "setenv(\"$key\", \"$value\", 0);"
+}
+
+# unsetEnv KEY
+unsetEnv() {
+    local key=$(escapeStringLiteral "$1")
+    printf "%s" "unsetenv(\"$key\");"
+}
+
+# Put the entire source code into const char* SOURCE_CODE to make it readable after compilation.
+# documentationString SOURCE_CODE
+documentationString() {
+    local docs=$(escapeStringLiteral $'\n----------\n// This binary wrapper was compiled from the following generated C-code:\n'"$1"$'\n----------\n')
+    printf "%s" "const char * SOURCE_CODE = \"$docs\";"
+}
+
+# Makes it safe to insert STRING within quotes in a C String Literal.
+# escapeStringLiteral STRING
+escapeStringLiteral() {
+    local result
+    result=${1//$'\\'/$'\\\\'}
+    result=${result//\"/'\"'}
+    result=${result//$'\n'/"\n"}
+    result=${result//$'\r'/"\r"}
+    printf "%s" "$result"
+}
+
+concat3Fn() {
+    printf "%s\n" 'char *concat3(char *x, char *y, char *z) {'
+    printf "%s\n" '    int xn = strlen(x);'
+    printf "%s\n" '    int yn = strlen(y);'
+    printf "%s\n" '    int zn = strlen(z);'
+    printf "%s\n" '    char *res = malloc(sizeof(*res)*(xn + yn + zn + 1));'
+    printf "%s\n" '    for (int i = 0; i < xn; ++i) res[i] = x[i];'
+    printf "%s\n" '    for (int i = 0; i < yn; ++i) res[xn+i] = y[i];'
+    printf "%s\n" '    for (int i = 0; i < zn; ++i) res[xn+yn+i] = z[i];'
+    printf "%s\n" "    res[xn+yn+zn] = '\0';"
+    printf "%s\n" '    return res;'
+    printf "%s\n" '}'
```

Comment:
Why not use multiline strings or EOF instead?

Reply:
The reason I've been using printf over echo is that it is supposedly more standardized/more portable (https://unix.stackexchange.com/questions/65803/why-is-printf-better-than-echo/65819#65819)

That said, the edge cases where this is true might not apply/make a difference here.

I've switched to using a multiline string instead of multiple printf-lines as you suggested